File: php/tests/smoke.php
<?php
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

declare(strict_types=1);

use Gmeow\Gts\Gts;
use Gmeow\Gts\GtsException;
use Gmeow\Gts\GtsStatus;

require __DIR__ . '/autoload.php';

if ($argc !== 4) {
    fwrite(
        STDERR,
        "usage: php -d ffi.enable=1 php/tests/smoke.php vectors/01-minimal.gts vectors/04-damaged-frame.gts vectors/28-empty-file.gts\n"
    );
    exit(2);
}

try {
    $gts = Gts::load();
    if ($gts->abiVersion() !== 1) {
        throw new RuntimeException(sprintf('Unexpected ABI version: %d', $gts->abiVersion()));
    }
    if ($gts->version() === '') {
        throw new RuntimeException('Empty library version.');
    }

    $input = file_get_contents($argv[1]);
    if ($input === false) {
        throw new RuntimeException(sprintf('Unable to read vector: %s', $argv[1]));
    }
    $damaged = file_get_contents($argv[2]);
    if ($damaged === false) {
        throw new RuntimeException(sprintf('Unable to read vector: %s', $argv[2]));
    }
    $empty = file_get_contents($argv[3]);
    if ($empty === false) {
        throw new RuntimeException(sprintf('Unable to read vector: %s', $argv[3]));
    }

    expectJsonProperty('build metadata', $gts->buildMetadataJson(), 'schema', 'gts-capi-build-v1');
    expectJsonProperty('capabilities', $gts->capabilitiesJson(), 'schema', 'gts-capi-capabilities-v1');
    $cleanRead = $gts->readJson($input);
    expectJsonProperty('php clean-read read JSON', $cleanRead, 'schema', 'gts-capi-read-v1');
    expectJsonProperty('php clean-read read JSON', $cleanRead, 'clean', true);
    expectJsonProperty('verify JSON', $gts->verifyJson($input), 'schema', 'gts-capi-verify-v1');

    $damagedRead = $gts->readJson($damaged);
    expectJsonProperty('php damaged-diagnostic-read read JSON', $damagedRead, 'schema', 'gts-capi-read-v1');
    expectJsonProperty('php damaged-diagnostic-read read JSON', $damagedRead, 'clean', false);
    expectDiagnostic('php damaged-diagnostic-read read JSON', $damagedRead, 'DamagedFrame');
    expectGtsException(
        'php damaged-diagnostic-read toNQuads',
        fn (): string => $gts->toNQuads($damaged),
        GtsStatus::DIAGNOSTIC
    );

    $emptyRead = $gts->readJson($empty);
    expectJsonProperty('php empty-malformed-refusal read JSON', $emptyRead, 'schema', 'gts-capi-read-v1');
    expectJsonProperty('php empty-malformed-refusal read JSON', $emptyRead, 'clean', false);
    expectDiagnostic('php empty-malformed-refusal read JSON', $emptyRead, 'EmptyFile');
    expectGtsException(
        'php empty-malformed-refusal toNQuads',
        fn (): string => $gts->toNQuads($empty),
        GtsStatus::DIAGNOSTIC
    );

    $nquads = $gts->toNQuads($input);
    expectContains('N-Quads', $nquads, '"Cat"@en');

    $roundTrip = $gts->fromNQuads($nquads);
    if ($roundTrip === '') {
        throw new RuntimeException('Round-trip GTS output was empty.');
    }

    expectGtsException(
        'php malformed-nquads-refusal fromNQuads',
        fn (): string => $gts->fromNQuads(getenv('GTS_WRAPPER_BAD_NQUADS') ?: "<https://example/s> <https://example/p> .\n"),
        GtsStatus::PARSE
    );

    $temp = sys_get_temp_dir() . '/gts-php-smoke-' . bin2hex(random_bytes(8));
    $sourceDir = $temp . '/src';
    $unpackDir = $temp . '/unpack';
    try {
        if (!mkdir($sourceDir, 0777, true) && !is_dir($sourceDir)) {
            throw new RuntimeException(sprintf('Unable to create %s.', $sourceDir));
        }
        file_put_contents($sourceDir . '/a.txt', "hello\n");

        $packed = $gts->filesPack([$sourceDir]);
        expectJsonProperty('files diff', $gts->filesDiffJson($packed, $sourceDir), 'clean', true);
        expectJsonProperty('files unpack', $gts->filesUnpack($packed, $unpackDir), 'ok', true);
        if (!is_file($unpackDir . '/a.txt')) {
            throw new RuntimeException('Unpacked file missing.');
        }
    } finally {
        removeTree($temp);
    }
} catch (Throwable $error) {
    fwrite(STDERR, $error . PHP_EOL);
    exit(1);
}

function expectJsonProperty(string $label, string $json, string $property, mixed $expected): void
{
    $decoded = json_decode($json, true, 512, JSON_THROW_ON_ERROR);
    if (!is_array($decoded) || !array_key_exists($property, $decoded)) {
        throw new RuntimeException(sprintf('%s missing JSON property %s.', $label, $property));
    }
    if ($decoded[$property] !== $expected) {
        throw new RuntimeException(sprintf('%s JSON property %s had an unexpected value.', $label, $property));
    }
}

function expectContains(string $label, string $haystack, string $needle): void
{
    if (!str_contains($haystack, $needle)) {
        throw new RuntimeException(sprintf('%s did not contain %s.', $label, $needle));
    }
}

function expectDiagnostic(string $label, string $json, string $expectedCode): void
{
    $decoded = json_decode($json, true, 512, JSON_THROW_ON_ERROR);
    if (!is_array($decoded) || !isset($decoded['diagnostics']) || !is_array($decoded['diagnostics'])) {
        throw new RuntimeException(sprintf('%s missing diagnostics array.', $label));
    }
    foreach ($decoded['diagnostics'] as $diagnostic) {
        if (is_array($diagnostic) && ($diagnostic['code'] ?? null) === $expectedCode) {
            return;
        }
    }
    throw new RuntimeException(sprintf('%s missing diagnostic %s.', $label, $expectedCode));
}

/**
 * @param callable(): mixed $callback
 */
function expectGtsException(string $label, callable $callback, int $expectedStatus): void
{
    try {
        $callback();
    } catch (GtsException $error) {
        if ($error->status !== $expectedStatus) {
            throw new RuntimeException(sprintf('%s expected %s, got %s.', $label, GtsStatus::name($expectedStatus), GtsStatus::name($error->status)));
        }
        if ($error->errorCode === '' || $error->detail === '') {
            throw new RuntimeException(sprintf('%s structured error did not include code and detail.', $label));
        }
        return;
    }
    throw new RuntimeException(sprintf('%s did not fail with %s.', $label, GtsStatus::name($expectedStatus)));
}

function removeTree(string $path): void
{
    if (!file_exists($path) && !is_link($path)) {
        return;
    }
    if (is_file($path) || is_link($path)) {
        unlink($path);
        return;
    }
    foreach (scandir($path) ?: [] as $entry) {
        if ($entry === '.' || $entry === '..') {
            continue;
        }
        removeTree($path . '/' . $entry);
    }
    rmdir($path);
}

File: ruby/tests/smoke.rb
# SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
# SPDX-License-Identifier: MIT OR Apache-2.0

require "fileutils"
require "json"
require "tmpdir"

unless ENV["GTS_RUBY_SMOKE_INSTALLED"] == "1"
  $LOAD_PATH.unshift(File.expand_path("../lib", __dir__))
end
require "gmeow/gts"

def expect_json_property(label, json, property, expected)
  decoded = JSON.parse(json)
  raise "#{label} missing JSON property #{property}." unless decoded.key?(property)
  return if decoded.fetch(property) == expected

  raise "#{label} JSON property #{property} had an unexpected value."
end

def expect_contains(label, haystack, needle)
  raise "#{label} did not contain #{needle}." unless haystack.include?(needle)
end

def expect_diagnostic(label, json, expected_code)
  decoded = JSON.parse(json)
  diagnostics = decoded.fetch("diagnostics")
  return if diagnostics.any? { |diagnostic| diagnostic.fetch("code") == expected_code }

  raise "#{label} missing diagnostic #{expected_code}."
end

def expect_gts_error(label, expected_status)
  yield
rescue Gmeow::Gts::Error => error
  raise "#{label} expected #{Gmeow::Gts.status_name(expected_status)}, got #{error.status_name}." unless error.status == expected_status
  raise "#{label} structured error did not include code and detail." if error.code.empty? || error.detail.empty?

  return
else
  raise "#{label} did not fail with #{Gmeow::Gts.status_name(expected_status)}."
end

unless ARGV.length == 3
  warn "usage: ruby -I ruby/lib ruby/tests/smoke.rb vectors/01-minimal.gts vectors/04-damaged-frame.gts vectors/28-empty-file.gts"
  exit 2
end

gts = Gmeow::Gts.load
raise "Unexpected ABI version: #{gts.abi_version}" unless gts.abi_version == Gmeow::Gts::ABI_VERSION
raise "Empty library version." if gts.version.empty?

input = File.binread(ARGV.fetch(0))
damaged = File.binread(ARGV.fetch(1))
empty = File.binread(ARGV.fetch(2))

expect_json_property("build metadata", gts.build_metadata_json, "schema", "gts-capi-build-v1")
expect_json_property("capabilities", gts.capabilities_json, "schema", "gts-capi-capabilities-v1")
clean_read = gts.read_json(input)
expect_json_property("ruby clean-read read JSON", clean_read, "schema", "gts-capi-read-v1")
expect_json_property("ruby clean-read read JSON", clean_read, "clean", true)
expect_json_property("verify JSON", gts.verify_json(input), "schema", "gts-capi-verify-v1")

damaged_read = gts.read_json(damaged)
expect_json_property("ruby damaged-diagnostic-read read JSON", damaged_read, "schema", "gts-capi-read-v1")
expect_json_property("ruby damaged-diagnostic-read read JSON", damaged_read, "clean", false)
expect_diagnostic("ruby damaged-diagnostic-read read JSON", damaged_read, "DamagedFrame")
expect_gts_error("ruby damaged-diagnostic-read to_nquads", Gmeow::Gts::Status::DIAGNOSTIC) do
  gts.to_nquads(damaged)
end

empty_read = gts.read_json(empty)
expect_json_property("ruby empty-malformed-refusal read JSON", empty_read, "schema", "gts-capi-read-v1")
expect_json_property("ruby empty-malformed-refusal read JSON", empty_read, "clean", false)
expect_diagnostic("ruby empty-malformed-refusal read JSON", empty_read, "EmptyFile")
expect_gts_error("ruby empty-malformed-refusal to_nquads", Gmeow::Gts::Status::DIAGNOSTIC) do
  gts.to_nquads(empty)
end

nquads = gts.to_nquads(input)
expect_contains("N-Quads", nquads, '"Cat"@en')

round_trip = gts.from_nquads(nquads)
raise "Round-trip GTS output was empty." if round_trip.empty?

expect_gts_error("ruby malformed-nquads-refusal from_nquads", Gmeow::Gts::Status::PARSE) do
  gts.from_nquads(ENV.fetch("GTS_WRAPPER_BAD_NQUADS", "<https://example/s> <https://example/p> .\n"))
end

Dir.mktmpdir("gts-ruby-smoke-") do |temp|
  source_dir = File.join(temp, "src")
  unpack_dir = File.join(temp, "unpack")

  FileUtils.mkdir_p(source_dir)
  File.binwrite(File.join(source_dir, "a.txt"), "hello\n")

  packed = gts.files_pack([source_dir])
  expect_json_property("files diff", gts.files_diff_json(packed, source_dir), "clean", true)
  expect_json_property("files unpack", gts.files_unpack(packed, unpack_dir), "ok", true)

  unpacked = File.join(unpack_dir, "a.txt")
  raise "Unpacked file missing." unless File.file?(unpacked)
  raise "Unpacked file content mismatch." unless File.binread(unpacked) == "hello\n"
end
